### ISSUE TITLE
Plugin using BouncyCastle isn't properly isolated

### ISSUE THREAD

**Issue by melix** [gradle/gradle#16605]
### Context

This is a follow-up to #16585. Plugins should be able to use their own dependencies, independently of the version shipped by Gradle for its own internal use. In general, [this works fine](https://github.com/gradle/gradle/blob/d798519c4f13f91fb6192cae3a87d51234f34955/subprojects/plugin-development/src/integTest/groovy/org/gradle/plugin/devel/impldeps/GradleImplDepsVisibilityIntegrationTest.groovy#L49-L73).

However, it seems that for BouncyCastle, isolation doesn't work properly. It _may_ be related to how we initialize BouncyCastle, or it may be that we don't filter out this dependency properly in the classloader hierarchy.

However, as illustrated in https://github.com/gradle/gradle/issues/16585#issuecomment-802900212 something doesn't quite work.

**Comment by aalmiray** [comment]
The problem lies in the following code found in `SecuritySupport`

```java
if (Security.getProvider(BouncyCastleProvider.PROVIDER_NAME) == null) {
    Security.setProperty("crypto.policy", "unlimited");
    Security.addProvider(new BouncyCastleProvider());
} 
```

Once could workaround this issue by inserting another BC provider ahead of the existing one

```java
Provider[] providers = Security.getProviders();
for (int i = 0; i < providers.length; i++) {
    if (providers[i].getName().equals(BouncyCastleProvider.PROVIDER_NAME)) {
        Security.insertProviderAt(new BouncyCastleProvider(), i);
        break;
    }
}
```

However classloading would still have happened by this time, thus an incompatible version of BouncyCastle is likely loaded. Consumers are required to shade BC and register it before. don't know what implications may be if Gradle were to shade BC.

**Comment by melix** [comment]
The workaround worked for you? FWIW, I understood that we couldn't really shade BC, because it's a signed jar and that if shaded then the security layer would fail loading.

**Comment by aalmiray** [comment]
Yes, the workaround appears to work for my case. Besides being a signed JAR one would have to contemplate export laws restrictions IIRC.

**Comment by aalmiray** [comment]
It's also possible to remove a provider (instead of inserting one with the same name ahead of the existing one) however I believe may still have classloading issues if the bundled BC dependency is found in the classpath before than the one provided by consumers. Also, what if the BC classes were loaded first?

**Comment by melix** [comment]
So there are 2 places where we use BC (except in tests):

- signing publications
- dependency verification

Signing could be refactored to execute in workers, I guess.
For dependency verification, it's more complicated since it happens during dependency resolution and that using workers here would be too expensive.

In any case, the problem is that the `Security` class is a top-level one, so this is the one which "leaks" BC. Would be great to figure out if we can avoid static initialization via `Security#addProvider` at all...

**Comment by aalmiray** [comment]
Would it be possible to defer `Security#addProvider` til the last possible moment?

**Comment by melix** [comment]
I guess we could but then the problem would be the other way around: the provider registered by the plugin would leak and dependency verification would fail, or signing would fail.

Also I suspect that because your plugin registers a new provider, then it leaks (or minimally the BC classes you use) to the system classloader via `Security#addProvider` and there's a memory leak.

**Comment by aalmiray** [comment]
Yeah, I suppose that once the shaded BC provider is added then anyone accessing it via name (i.e, "BC") will get the shaded version. Given that the shaded BC has its classes relocated to a prefixed package then that would not cause a problem _IF_ other plugins directly refer to non-shaded BC classes. Accessing BC via `Security` remains a problem either way.

**Comment by dsvensson** [comment]
Just tried to create a plugin under buildSrc that updates an AWS Lambda function, and it fails with `Caused by: java.lang.NoSuchFieldError: DOCUMENT` due to `MarshallingType` being found in org.gradle.internal.impldep.com.amazonaws.protocol, so this problem is more widespread than BouncyCastle.
